[dnf5] Improve performance of ifilter_installed() and ifilter_available()

jrohel opened the pull request:
The code adds/removes only installed packages.
Optimization: The cycle goes through only part of the pool from `installed_repo->start` to `installed_repo->end`.

Requires https://github.com/rpm-software-management/libdnf/pull/1092

jrohel committed:
Improve performance of ifilter_installed() and ifilter_available()

Idea: It is enought to iterate only the part of the pool that contains
the installed packages.

We can start the iteration from "installed_repo->start".
We can test current iterator value against "installed_repo->end" to stop
the iteration behind the installed repository.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -1790,10 +1790,11 @@ SolvQuery & SolvQuery::ifilter_installed() {
-    for (PackageId candidate_id : p_impl->query_result) {
-        Solvable * solvable = solv::get_solvable(pool, candidate_id);
+    for (auto candidate_id = installed_repo->start; candidate_id < installed_repo->end; ++candidate_id) {
+        auto pkg_candidate_id = PackageId(candidate_id);
```
I suggest that it is not right approach, but definitely good direction. Between `installed_repo->start` and `candidate_id < installed_repo->end` there could be Ids that are not packages. In theory it should be not harmful.
In case that `p_impl->query_result` is empty or with few elements, the new logic will be slower.
What about to iterate over `p_impl->query_result` with interval between `installed_repo->start`; `candidate_id < installed_repo->end`?

jrohel committed:
Improve performance of ifilter_installed()

Result is in "filter_result". There is no need for &=.

Note: The optimization replaces the original "query_result" buffer.
Is it OK? Or we need to use original slower code.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -1794,9 +1794,13 @@ SolvQuery & SolvQuery::ifilter_installed() {
         Solvable * solvable = solv::get_solvable(pool, candidate_id);
         if (solvable->repo == installed_repo) {
             filter_result.add_unsafe(candidate_id);
+            continue;
```
`for (PackageId candidate_id : p_impl->query_result) {`
It would be also nice if we can start to iterate from `installed_repo->start`

jrohel replied:
Yes, that would be nice. But in my commit message is explanation:
Theoretically, the iteration can start from "installed_repo->start",
**but we can't set iterator to this position. The iterator does not have
a setter only an operator ++. -> Iteration must start from the beginning
of the pool.** There is usually no slowdown because the installed repository
is usually the first repository in the pool.

But I have already solved the problem. I will prepare a commit.

jrohel replied:
I added `jump()` method into `rpm::solv::SolvMapIterator` in PR https://github.com/rpm-software-management/libdnf/pull/1092 .
And the current version of the optimization uses it to start the iteration from `installed_repo-> start`.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -1794,9 +1794,13 @@ SolvQuery & SolvQuery::ifilter_installed() {
         }
     }
-    p_impl->query_result &= filter_result;
+    p_impl->query_result = std::move(filter_result);
```
I know that it is faster but in case that we will give const reference to `query_result` as API, we have to document it.

jrohel replied:
To be honest, I'm not sure if we want this optimization. I prepared a comit to know your opinion.

jrohel replied:
I added TODO.

j-mracek commented:
LGTM
